### PR TITLE
drivers: sensor: Add support for Broadcom APDS-9306

### LINKED ISSUE #74546: Add support for Broadcom APDS9306
**Is your enhancement proposal related to a problem? Please describe.**
The driver for the sensor module is missing.

**Describe the solution you'd like**
I prepare a PR to add the driver for the module from our Smartwatch project (see [here](https://github.com/jakkra/ZSWatch/tree/main/app/drivers/sensor/apds9306)).

**Additional context**
[APDS9306](https://www.broadcom.com/products/optical-sensors/ambient-light-photo-sensors/apds-9306-065)

### PR DESCRIPTION
**Pull request by Kampi**
**Description**

This PR adds support for [Broadcom / Avago APDS-9306](https://www.broadcom.com/products/optical-sensors/ambient-light-photo-sensors/apds-9306-065) ambient light sensor connected to the I2C bus.

**Supported functions**

- [x] Setting of sampling frequency, gain and resolution
- [x] Reading of sensor values

**Tested**

- [x] Unit test
- [x] Hardware

Closes #74546

### CODE REVIEW

**Comment by Kampi** [comment]
Open questions:

The driver is using `SENSOR_APDS9306_ATTR_GAIN` and `SENSOR_APDS9306_ATTR_RESOLUTION` as private attributes to set and get the resolution and the gain. Imo these attributes are generic and valid for other sensors, so I would like to add them to the sensor subsystem in `sensor.h`

File: drivers/sensor/apds9306/Kconfig [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+# Copyright (c) 2024 Daniel Kampert
+# Author: Daniel Kampert <DanielKampert@Kampis-Elektroecke.de>
+
+menuconfig APDS9306
+	bool "APDS9306 Sensor"
+	default y
+	depends on DT_HAS_AVAGO_APDS9306_ENABLED
+	select I2C
+	help
+		Enable the driver for the APDS9306 digital light sensor.
```
Comment (MaureenHelm) [inline drivers/sensor/apds9306/Kconfig:11]:
```suggestion
	  Enable the driver for the APDS9306 digital light sensor.
```

File: drivers/sensor/apds9306/apds9306.c [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+		mask = 0x07 << 0x00;
+		temp = value->val1 & 0x07;
+	} else if (attribute == SENSOR_APDS9306_ATTR_GAIN) {
+		reg = APDS9306_REGISTER_ALS_GAIN;
+		mask = 0x07 << 0x00;
+		temp = value->val1 & 0x07;
+	} else if (attribute == SENSOR_APDS9306_ATTR_RESOLUTION) {
+		reg = APDS9306_REGISTER_ALS_MEAS_RATE;
+		mask = 0x07 << 0x04;
+		temp = (value->val1 & 0x07) << 0x04;
```
Comment (MaureenHelm) [inline drivers/sensor/apds9306/apds9306.c:160]: Look at using `GENMASK` and `FIELD_PREP` macros from `include/zephyr/sys/util.h`

File: drivers/sensor/apds9306/apds9306.c [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+}
+
+static int apds9306_init(const struct device *dev)
+{
+	uint8_t value;
+	const struct apds9306_config *config = dev->config;
+
+	LOG_DBG("Start to initialize APDS9306...");
+
+	if (!device_is_ready(config->i2c.bus)) {
```
Comment (MaureenHelm) [inline drivers/sensor/apds9306/apds9306.c:328]: Use `i2c_is_ready_dt`

File: drivers/sensor/apds9306/apds9306.c [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+		break;
+	}
+	default: {
+		return -ENOTSUP;
+	}
+	}
+
+	return 0;
+}
+#endif
```
Comment (MaureenHelm) [inline drivers/sensor/apds9306/apds9306.c:373]: This doesn't do anything, remove.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Add this sensor to `tests/drivers/build_all/sensor`

> The driver is using SENSOR_APDS9306_ATTR_GAIN and SENSOR_APDS9306_ATTR_RESOLUTION as private attributes to set and get the resolution and the gain. Imo these attributes are generic and valid for other sensors, so I would like to add them to the sensor subsystem in sensor.h

Not opposed to this, but there are a couple of other sensors that define similar private attributes so let's look at introducing common attributes in a separate PR.

**Comment by Kampi** [comment]
@MaureenHelm would it be okay to do add this attributes at first and then add the sensor? So the sensor driver doesn't need a change after both PR were merged.

**Comment by Kampi** [comment]
Replaced the custom attributes with the new `SENSOR_ATTR_GAIN` and `SENSOR_ATTR_RESOLUTION` attributes.

File: tests/drivers/build_all/sensor/i2c.dtsi [hunk 5]
```diff
@@ -559,6 +559,15 @@ test_i2c_vcnl4040: vcnl4040@51 {
...
 };
 
+apds_9306: apds9306@52 {
+	compatible = "avago,apds9306";
+	reg = <0x52>;
+	status = "okay";
+	gain = <1>;
+	resolution = <13>;
+	frequency = <2000>;
+};
```
Comment (MaureenHelm) [inline tests/drivers/build_all/sensor/i2c.dtsi:569]: Move to the bottom of the file and increment the address

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: tests/drivers/build_all/sensor/i2c.dtsi (hunk 5 above)
Comment (Kampi) [reply tests/drivers/build_all/sensor/i2c.dtsi:569]: So I should pick the next free address, regardless if this is the real address or not?

**Review by MaureenHelm — APPROVED** [review]

**Comment by Kampi** [comment]
@MaureenHelm can you please tell me how do we continue here?

**Review by MaureenHelm — APPROVED** [review]

**Review by ubieda — APPROVED** [review]

**Pull request merged**